Allow arbitrary configuration options to be overridden via the CLI (#9599)
astral-sh/ruff: docs/configuration.md

## Command-line interface

> All configuration options can be provided or overridden via the command-line.
> Some of the most common have their own dedicated command-line flag.
> This includes the configuration options relating to rule enablement and disablement,
> file discovery, and logging level:
> 
> ```shell
> ruff check path/to/code/ --select F401 --select F403 --quiet
> ```
> 
> Configuration options that do not have a dedicated flag can be set or
> overridden using the `--config` flag. The `--config` flag is most often
> used to point to the configuration file that you would like ruff to use,

Reviewer: Nit: I would explain one concept at a time by

1. Explain the configuration options that can be overriden globally (they also have different semantics)
2. Explain the `--config` option

```suggestion
Some configuration options can be provided or overriden via the command-line, such as those related to rule enablement and disablement, file discovery, logging level, and more:

```shell
ruff check path/to/code/ --select F401 --select F403 --quiet
```

Configuration options that do not have a dedicated flag can be set using the `--config` flag. The `--config` flag is most often
used to point to the configuration file that you would like ruff to use,
```

I think it would be worth explaining in more detail how the config options are overriden. Does `--config line-length=100` override the `line-length` settings for all files or do configuration files in sub-directories have priority (because they're closer to the file)?

## Command-line interface

Some configuration options can be provided or overridden via dedicated flags on the command line.
This includes those related to rule enablement and disablement,
file discovery, logging level, and more:

```shell
ruff check path/to/code/ --select F401 --select F403 --quiet
```

> ### The `--config` CLI flag

Author: Something like this?

```suggestion
One flag that deserves special mention is the `--config` flag.

### The `--config` CLI flag
```

> ### The `--config` CLI flag

Reviewer 2: Yeah, or even like... "For everything else, there's the `--config` flag."

> ### The `--config` CLI flag

Author: Added in 789d1d0d17e569f4a9ac96be3473af2fa0f66b25 👍

The `--config` flag has two uses. It is most often used to point to the
configuration file that you would like ruff to use, for example:

```shell
ruff check path/to/directory --config path/to/ruff.toml
```

However, the `--config` flag can also be used to provide arbitrary
overrides of configuration settings using TOML `<KEY> = <VALUE>` pairs.
This is mostly useful in situations where you wish to override a configuration setting
that does not have a dedicated command-line flag.

In the below example, the `--config` flag is the only way of overriding the
`dummy-variable-rgx` configuration setting from the command line,
since this setting has no dedicated CLI flag. The `per-file-ignores` setting
could also have been overridden via the `--per-file-ignores` dedicated flag,
but using `--config` to override the setting is also fine:

```shell
ruff check path/to/file --config path/to/ruff.toml --config "lint.dummy-variable-rgx = '__.*'" --config "lint.per-file-ignores = {'some_file.py' = ['F841']}"
```

Configuration options passed to `--config` are parsed in the same way
as configuration options in a `ruff.toml` file.

> As such, options specific to the ruff linter need to be prefixed with `lint.`

Reviewer 2: Nit: capitalize "Ruff" here.

(`--config "lint.dummy-variable-rgx = '__.*'"` rather than simply

> `--config "dummy-variable-rgx = '__.*'"`), and options specific to the ruff formatter

Reviewer 2: Nit: capitalize "Ruff" here.

need to be prefixed with `format.`.

If a specific configuration option is simultaneously overridden by
a dedicated flag and by the `--config` flag, the dedicated flag

> always takes priority. In this example, the maximum permitted line

Reviewer 2: Nit: can omit "always" here.

length will be set to 90, not 100:

```shell
ruff format path/to/file --line-length=90 --config "line-length=100"
```

Specifying `--config "line-length=90"` will override the `line-length`

> setting from *all* configuration files detected by ruff,

Reviewer 2: Nit: capitalize "Ruff"
